### ISSUE TITLE
Loss increases drastically at beginning of new epoch on custom dataset

### ISSUE THREAD

**Issue by arkilpatel** [pytorch/torchtune#1714]
Hi, I am trying to finetune models (llama3.1, qwen2, etc) 7B versions on my custom math reasoning dataset. My data is of size 3800 and formatted in json with "input" and "output" fields.

I am seeing the loss increase drastically at the beginning of new epochs:

![W B Chart 9_29_2024, 6_14_22 PM](https://github.com/user-attachments/assets/e0b53263-0d0e-44f9-9531-990d22a196a2)

This coincides with drastic decrease in tokens per second per gpu (idk if this is relevant, just a pattern I observed):

![W B Chart 9_29_2024, 6_14_34 PM](https://github.com/user-attachments/assets/cae6396a-0831-4f1a-af8b-46f9e08357d6)

My config:

```
tokenizer:
  _component_: torchtune.models.qwen2.qwen2_tokenizer
  path: tmp/qwen/vocab.json
  merges_file: tmp/qwen/merges.txt
  max_seq_len: null

# Model Arguments
model:
  _component_: torchtune.models.qwen2.lora_qwen2_7b
  lora_attn_modules: ['q_proj', 'v_proj', 'k_proj', 'output_proj']
  apply_lora_to_mlp: True
  apply_lora_to_output: True
  lora_rank: 8
  lora_alpha: 16
  lora_dropout: 0.0

checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: tmp/qwen/
  checkpoint_files: [
    model-00001-of-00004.safetensors,
    model-00002-of-00004.safetensors,
    model-00003-of-00004.safetensors,
    model-00004-of-00004.safetensors
  ]
  recipe_checkpoint: null
  output_dir: tmp/qwen/
  model_type: QWEN2
resume_from_checkpoint: False
save_adapter_weights_only: False

# Dataset and Sampler
dataset:
  _component_: torchtune.datasets.instruct_dataset
  source: json
  split: train
  train_on_input: False
  data_files: data/synmath_train.json
seed: null
shuffle: True
batch_size: 4

# Optimizer and Scheduler
optimizer:
  _component_: torch.optim.AdamW
  fused: True
  weight_decay: 0.01
  lr: 3e-4
lr_scheduler:
  _component_: torchtune.modules.get_cosine_schedule_with_warmup
  num_warmup_steps: 100

loss:
  _component_: torchtune.modules.loss.CEWithChunkedOutputLoss

# Training
epochs: 3
max_steps_per_epoch: null
gradient_accumulation_steps: 32

# Environment
device: cuda
dtype: bf16
enable_activation_checkpointing: False
```

This pattern holds across multiple models I tried: llama, mistral, qwen. Is there something I am doing wrong?

**Comment by SalmanMohammadi** [comment]
Hey @arkilpatel. Thanks for trying out torchtune! Could you confirm whether this is happening at the start of every epoch, or at the end of every epoch? If this is happening for the last batch in every epoch? If the latter, see my response below. 

One thing I think might be contributing to this phenomenon is the values you're using for `batch_size` and `gradient_acummulation_steps`. You are currently using `4 * 32 = 128` samples every time you perform an optimizer step on your model weights, which could be quite a large portion of your dataset! It looks like you're only doing about ~7 gradient updates per epoch.

Now, we log the loss and memory statistics every gradient update step. However, it looks like the last gradient update is significantly smaller (around half of the rest of your updates), so this could explain the shapes of your graphs: toks/sec dips because there are less samples in the batch for your GPU to take advantage of, and the loss peaks because you're inducing more variance in your gradient updates by using fewer samples.

Overall it's normal to expect some variance when training your models. It looks like the global trend is that your loss is decreasing which is what we'd expect, so these graphs aren't necessarily cause for concern.

One thing I'd reccomend trying out is adding `packed=True` to your dataset config:
```
dataset:
  _component_: torchtune.datasets.instruct_dataset
  source: json
  split: train
  train_on_input: False
  data_files: data/synmath_train.json

**Comment by SalmanMohammadi** [comment]
Hey @arkilpatel. Thanks for trying out torchtune! Could you confirm whether this is happening at the start of every epoch, or at the end of every epoch? If this is happening for the last batch in every epoch? If the latter, see my response below. 

One thing I think might be contributing to this phenomenon is the values you're using for `batch_size` and `gradient_acummulation_steps`. You are currently using `4 * 32 = 128` samples every time you perform an optimizer step on your model weights, which could be quite a large portion of your dataset! It looks like you're only doing about ~7 gradient updates per epoch.

Now, we log the loss and memory statistics every gradient update step. However, it looks like the last gradient update is significantly smaller (around half of the rest of your updates), so this could explain the shapes of your graphs: toks/sec dips because there are less samples in the batch for your GPU to take advantage of, and the loss peaks because you're inducing more variance in your gradient updates by using fewer samples.

Overall it's normal to expect some variance when training your models. It looks like the global trend is that your loss is decreasing which is what we'd expect, so these graphs aren't necessarily cause for concern.

A couple things I'd reccomend trying out to generally improve performance is adding `packed=True` to your dataset config:
```
dataset:
  _component_: torchtune.datasets.instruct_dataset
  source: json
  split: train
  train_on_input: False
  data_files: data/synmath_train.json
  packed=True
```
Which should mean you consistently see more samples for every gradient update, and also boost training performance! Note: you may need to install the nightly version of torchtune to make sure this works (as we add `drop_last=True` to our DataLoader to prevent ragged batches. This will also mitigate the effect of a smaller final-batch I described above without `packed=True`).

I also think it's worth reducing `num_warmup_steps` for your `lr_scheduler` - in the graph above you're only doing about 40 steps in total. Lastly, try playing around with `lora_alpha` and `lora_rank` (see our docs for more information on how these affect training https://pytorch.org/torchtune/main/tutorials/memory_optimizations.html#parameter-efficient-fine-tuning-peft) if you're not seeing results as expected.

Lastly, make sure you have a chance to evaluate your trained model for the task you're hoping to use it for - you can try out our `generate_v2` recipe to make sure generations are as expected, or our `eleuther_eval` recipe to benchmark against well-documented evaluation tasks.

**Comment by arkilpatel** [comment]
Hi @SalmanMohammadi decreasing the `gradient_acummulation_steps` worked! I had completely overlooked that parameter so thanks for pointing it out!
